src/Microsoft.Kiota.Cli.Commons/Http/NativeHttpHeadersHandler.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Kiota.Cli.Commons.Http.Headers;
+
+namespace Microsoft.Kiota.Cli.Commons.Http;
+
+/// <summary>
+/// A <see cref="DelegatingHandler"/> that calls a function which should return
+/// a store (<see cref="IHeadersStore"/>) instance and adds the stored headers
+/// to any request before sending it.
+/// </summary>
+public class NativeHttpHeadersHandler : DelegatingHandler
+{
+    private const string AddHeaderWarningTpl = "Could not add the {Kind}header {HeaderName}";
+
+    private readonly Func<IHeadersStore> _headersStoreGetter;
+    private readonly ILogger<NativeHttpHeadersHandler>? _logger;
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="NativeHttpHeadersHandler"/>
+    /// that will call the provided <paramref name="headersStoreGetter"/> function
+    /// and attempt to add all the headers in the returned
+    /// <see cref="IHeadersStore"/> instance. This handler calls the
+    /// <see cref="IHeadersStore.Drain"/> function on the instance.
+    /// </summary>
+    /// <param name="headersStoreGetter">
+    /// A function that when called, should return a <see cref="IHeadersStore"/>
+    /// instance containing headers to send.
+    /// </param>
+    /// <param name="logger">
+    /// An optional logger which defaults to null.
+    /// </param>
+    /// <remarks>
+    /// The class will not log any message if a logger is not provided.
+    /// </remarks>
+    public NativeHttpHeadersHandler(Func<IHeadersStore> headersStoreGetter,
+        ILogger<NativeHttpHeadersHandler>? logger = null)
+    {
+        _headersStoreGetter = headersStoreGetter;
+        _logger = logger;
+    }
+
+    /// <inheritdoc />
+    /// <remarks>
+    /// Adds headers to the <paramref name="request"/> before sending it.
+    /// </remarks>
+    /// <exception cref="FormatException">
+    /// The header name format is invalid. -or- New line characters in header values must be followed by a white-space
+    /// character.
+    /// </exception>
+    protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+        CancellationToken cancellationToken)
+    {
+        foreach (var headerItem in _headersStoreGetter().Drain())
+        {
+            try
+            {
+                // content headers defined in: https://www.rfc-editor.org/rfc/rfc2616
+                if (headerItem.Key is "Content-Encoding" or "Content-Language" or "Content-Length" or "Content-Location"
```

Comment:
move those constants to a hashset initialized with StringComparer.OrdinalIgnore case to:
1. get o(1) comparison instead of o(n)
2. add tolerance for different casing

Reply:
I've run a benchmark and the pattern matching is indeed faster than using a HashSet.

|                 Method |      Mean |     Error |    StdDev |
|----------------------- |----------:|----------:|----------:|
|   IsContentHeaderInSet | 11.650 ns | 0.6050 ns | 1.7839 ns |
| IsContentHeaderPattern |  3.277 ns | 0.1606 ns | 0.4661 ns |

Reply:
are you initializing the HashSet outside of the measured time?
Also this doesn't solve for casing

Reply:
> are you initializing the HashSet outside of the measured time?
I am

Reply:
I've added case insensitive comparison. It's a little slower, but not by much. I already tried having the headers in an array and looping with equality, but that was only 1ns faster than the HashSet. I think array lookup has some hidden overhead I didn't realize. could be processor cache differences.

|                      Method |      Mean |     Error |   StdDev |    Median | Allocated |
|---------------------------- |----------:|----------:|---------:|----------:|----------:|
|        IsContentHeaderInSet | 36.161 ns | 1.7263 ns | 5.063 ns | 35.795 ns |         - |
| IsContentHeaderEqualsManual |  6.416 ns | 0.4229 ns | 1.227 ns |  6.050 ns |         - |